### ISSUE TITLE
Yielding issue from nested call: apply fails, bind works in Duktape.Thread

### ISSUE THREAD

**Issue by kungfooman** [svaarala/duktape#1420]
Hey, I tried to make the example as simple as possible. apply is really strange, it seems to kill the coroutine immediatly (the coroutine will not even return undefined anymore, it just ends):

	function yieldvalue(value)
	{
	  Duktape.Thread.yield( value );
	}

	function testfunc() {
	  yieldvalue(111);
	  yieldvalue(222);
	  yieldvalue(333);

	  if (1) {
		yieldvalue.bind(123)(444); // works
	  } else {
		log("Hurr, will it work?!\n");
		yieldvalue.apply(123, [444]) // fails
	  }
	}

	thread = new Duktape.Thread(testfunc);

	function resumethread() {
	  try {
		var gotvalue = Duktape.Thread.resume(thread, [1,2,3]);
		log("got value: " + gotvalue + "\n");
	  } catch (e) {
		log("Thread ended!\n");
	  }
	}

	resumethread(); // got value: 111
	resumethread(); // got value: 222
	resumethread(); // got value: 333
	resumethread(); // got value: 444          OR: Thread ended
	resumethread(); // got value: undefined    OR: Thread ended
	resumethread(); // Thread ended!           OR: Thread ended

Basically, when it works, it prints:

	got value: 111
	got value: 222
	got value: 333
	got value: 444
	got value: undefined
	Thread ended!

But when using if(0) so the apply is used, the output is:

	got value: 111
	got value: 222
	got value: 333
	Hurr, will it work?!
	Thread ended!
	Thread ended!
	Thread ended!

I wonder why bind works, but apply does not. Isn't it basically exactly the same? I kinda need the "apply" to work, because it would allow me to write a super dynamic mega awesome game engine multiplayer scripting (which I will release later with MIT license of course :D).

**Comment by svaarala** [comment]
Current coroutine limitations include .call() and .apply(), see: http://duktape.org/guide.html#coroutines. The .call() / .apply() limitation will be removed in Duktape 2.1.

The .call() / .apply() limitations comes from an internal technical detail: yield is not allowed if there's a native call in the call stack of the yielding thread (if there is, we can't recreate the stack portably later). In Duktape 2.0 and prior .call() and .apply() are visible in the call stack so they prevent a yield.

To avoid the limitations one approach you can use right now is duk_suspend() and duk_resume(). You can combine that with native threads or native C stack switching (see e.g. https://github.com/rhoot/sc).

**Comment by fatcerberus** [comment]
In the general case, could `yield()` maybe throw an error in case the yield is not possible instead of just silently killing the coroutine?

**Comment by svaarala** [comment]
It does throw an error, but it's eaten by the example:

```
TypeError: invalid state
```

**Comment by fatcerberus** [comment]
Ah, I didn't see the try-catch in the example.  My bad :-)

**Comment by fatcerberus** [comment]
@kungfooman Since `.bind()` does work as shown by your example code, you can use it as a workaround to fake `.apply()` in the meantime until #1421.  Just do a bind on the target and then immediately call the resulting bound function.

**Comment by kungfooman** [comment]
@fatcerberus Yea, what I do atm is this:

    func.bind(self)(args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], args[8], args[9]);

That doesn't cause any errors, thanks to JS being so dynamic. It's just somewhat unnice. But glad to hear that the problem is known and will be fixed in 2.1 :)

Basically in the system I develop I have no clue beforehand how many arguments will be passed, hence I was in need for `.apply()`.

**Comment by svaarala** [comment]
@kungfooman It's not so much a bug but a known limitation :) There'a a few other cases on the prohibited constructs list which are unfortunate but difficult to fix -- like yielding from a .forEach(). The technical problem is that .forEach() is a native function, and to yield/resume properly there would need to be a way to reconstitute its state when resuming a thread. In practice this would mean rebuilding a C native stack entry and remembering the state of the iteration. All the limitations basically stem from this root cause.

Rebuilding a native C stack entry is possible, but it needs cooperation from each native function supporting a yield so it requires quite a lot of effort to implement fully across all the bindings (anything that invokes any function that might yield would need to do it). To also work for application native functions also the application would need to understand and deal with this, which gets quite awkward quickly IMO.

So I think the status in the medium term will be that:
- There are some yield() limitations (hopefully less and less as time goes on).
- duk_suspend()/duk_resume() and a native thread switch (either multiple threads or single thread and C stack switching) allows a much more comprehensive solution which has very few limitations (for example, cannot duk_suspend() from a finalizer) for either Duktape or application native calls.

The latter technique is very useful but there isn't a Wiki Howto yet on the subject. It would be nice to have a ready-to-use example for it.

**Comment by fatcerberus** [comment]
@kungfooman If you want to pass an array of arguments (as I assume that's why you're using `.apply` in the first place, you can shorten it a lot by:
```js
func.bind.apply(self, args)();
```

That will work even with the limitation because you're just doing the apply on `.bind()`.  The function itself is then called normally.

**Comment by svaarala** [comment]
@kungfooman The .call/.apply change didn't make it to 2.1 but there's a working prototype now in #1421 which should make its way into 2.2.

**Comment by svaarala** [comment]
There are now changes in master to allow yield from:
- new Foo() and Reflect.construct() calls
- .call(), .apply(), and Reflect.apply() calls

Closing this issue.
